x-pack/plugins/data_enhanced/server/search/session/session_service.ts

```diff
@@ -0,0 +1,363 @@
+    const updatedSessions = activeMappingObjects
+      ?.filter((so) => !so.error)
+      .map((sessionSavedObject) => {
+        const sessionInfo = this.sessionSearchMap.get(sessionSavedObject.id);
+        const idMapping = sessionInfo ? Object.fromEntries(sessionInfo.ids.entries()) : {};
+        sessionSavedObject.attributes.idMapping = {
+          ...sessionSavedObject.attributes.idMapping,
+          ...idMapping,
+        };
+        return sessionSavedObject;
+      });
+
+    const updateResults = await this.internalSavedObjectsClient.bulkUpdate<BackgroundSessionSavedObjectAttributes>(
+      updatedSessions
+    );
+    return updateResults.saved_objects;
+  }
+
+  public search<Request extends IKibanaSearchRequest, Response extends IKibanaSearchResponse>(
+    strategy: ISearchStrategy<Request, Response>,
+    searchRequest: Request,
+    options: ISearchOptions,
+    searchDeps: SearchStrategyDependencies,
+    deps: BackgroundSessionDependencies
+  ): Observable<Response> {
+    // If this is a restored background search session, look up the ID using the provided sessionId
+    const getSearchRequest = async () =>
+      !options.isRestore || searchRequest.id
+        ? searchRequest
+        : {
+            ...searchRequest,
+            id: await this.getId(searchRequest, options, deps),
+          };
+
+    return from(getSearchRequest()).pipe(
+      switchMap((request) => strategy.search(request, options, searchDeps)),
+      tapFirst((response) => {
+        if (searchRequest.id || !options.sessionId || !response.id || options.isRestore) return;
+        this.trackId(searchRequest, response.id, options, deps);
+      })
+    );
+  }
+
+  // TODO: Generate the `userId` from the realm type/realm name/username
+  public save = async (
+    sessionId: string,
+    {
+      name,
+      appId,
+      created = new Date().toISOString(),
+      expires = new Date(Date.now() + DEFAULT_EXPIRATION).toISOString(),
+      status = BackgroundSessionStatus.IN_PROGRESS,
+      urlGeneratorId,
+      initialState = {},
+      restoreState = {},
+    }: Partial<BackgroundSessionSavedObjectAttributes>,
+    { savedObjectsClient }: BackgroundSessionDependencies
+  ) => {
+    if (!name) throw new Error('Name is required');
+    if (!appId) throw new Error('AppId is required');
+    if (!urlGeneratorId) throw new Error('UrlGeneratorId is required');
+
+    this.logger.debug(`save | ${sessionId}`);
+
+    const attributes = {
+      name,
+      created,
+      expires,
+      status,
+      initialState,
+      restoreState,
+      idMapping: {},
+      urlGeneratorId,
+      appId,
+      sessionId,
+    };
+    const session = await savedObjectsClient.create<BackgroundSessionSavedObjectAttributes>(
+      BACKGROUND_SESSION_TYPE,
+      attributes,
+      { id: sessionId }
+    );
+
+    return session;
+  };
+
+  // TODO: Throw an error if this session doesn't belong to this user
```

Comment:
Not sure what your timeline looks like, but we are starting work on [OLS Phase 1](https://github.com/elastic/kibana/issues/82725), which might prove useful here.

Reply:
@legrego We have a PR up already, would you mind taking a look? https://github.com/elastic/kibana/pull/84975